File: scripts/fix-missing-alt-text.js
/**
 * Bulk Fix Missing Alt Text for Shopify Product Images
 * 
 * Usage:
 *   # Dry run (default):
 *   SHOPIFY_STORE_DOMAIN=inkitchonline.myshopify.com SHOPIFY_ADMIN_TOKEN=shpat_xxx node scripts/fix-missing-alt-text.js
 * 
 *   # Execute live updates:
 *   SHOPIFY_STORE_DOMAIN=inkitchonline.myshopify.com SHOPIFY_ADMIN_TOKEN=shpat_xxx node scripts/fix-missing-alt-text.js --commit
 */

const fs = require('fs');
const path = require('path');

// Configuration from environment or CLI arguments
const STORE_DOMAIN = process.env.SHOPIFY_STORE_DOMAIN || 'inkitchonline.myshopify.com';
const ADMIN_TOKEN = process.env.SHOPIFY_ADMIN_TOKEN || '';
const API_VERSION = process.env.SHOPIFY_API_VERSION || '2025-01';
const IS_DRY_RUN = !process.argv.includes('--commit') && process.env.DRY_RUN !== 'false';

const COLOR_OPTION_REGEX = /^(colou?r|farbe|couleur|coloris|colore|cor|shade|finish|style)$/i;

/**
 * Helper to make rate-limited & auto-retrying fetch calls to Shopify REST API
 */
async function shopifyRequest(endpointUrl, options = {}, retries = 5) {
  const url = endpointUrl.startsWith('http')
    ? endpointUrl
    : `https://${STORE_DOMAIN}/admin/api/${API_VERSION}${endpointUrl}`;

  const headers = {
    'Content-Type': 'application/json',
    'X-Shopify-Access-Token': ADMIN_TOKEN,
    ...options.headers,
  };

  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      const res = await fetch(url, { ...options, headers });

      if (res.status === 429) {
        const retryAfter = parseFloat(res.headers.get('Retry-After') || '2.0');
        console.warn(` Rate limited (429). Waiting ${retryAfter}s (attempt ${attempt}/${retries})...`);
        await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000 + 500));
        continue;
      }

      if (!res.ok) {
        const errorText = await res.text();
        throw new Error(`Shopify API error (${res.status} ${res.statusText}): ${errorText}`);
      }

      const linkHeader = res.headers.get('Link');
      const data = await res.json();
      return { data, linkHeader };
    } catch (err) {
      if (attempt === retries) throw err;
      const backoff = Math.pow(2, attempt) * 500;
      console.warn(` Network error: ${err.message}. Retrying in ${backoff}ms...`);
      await new Promise((resolve) => setTimeout(resolve, backoff));
    }
  }
}

/**
 * Parses pagination Link header to find the next page URL
 */
function getNextPageUrl(linkHeader) {
  if (!linkHeader) return null;
  const links = linkHeader.split(',');
  for (const link of links) {
    const match = link.match(/<([^>]+)>;\s*rel="next"/i);
    if (match) return match[1];
  }
  return null;
}

/**
 * Main Execution Function
 */
async function main() {
  console.log('====================================================');
  console.log('   SHOPIFY BULK ALT-TEXT FIXER');
  console.log('====================================================');
  console.log(` Store Domain : ${STORE_DOMAIN}`);
  console.log(` API Version  : ${API_VERSION}`);
  console.log(` Mode         : ${IS_DRY_RUN ? 'DRY RUN (No changes will be saved)' : 'COMMIT (Updating Shopify Admin live!)'}`);
  console.log('====================================================\n');

  if (!ADMIN_TOKEN) {
    console.error('ERROR: SHOPIFY_ADMIN_TOKEN environment variable is missing.');
    console.error('Please set SHOPIFY_ADMIN_TOKEN and run again.\n');
    console.error('Example:');
    console.error('  SHOPIFY_ADMIN_TOKEN=shpat_xxxx node scripts/fix-missing-alt-text.js --commit\n');
    process.exit(1);
  }

  let productsScanned = 0;
  let totalImagesChecked = 0;
  let skippedAlreadyHasAlt = 0;
  let updatedImagesCount = 0;
  const actionLog = [];

  let nextUrl = `/products.json?limit=250`;

  while (nextUrl) {
    console.log(`Fetching products page...`);
    const { data, linkHeader } = await shopifyRequest(nextUrl);
    const products = data.products || [];

    for (const product of products) {
      productsScanned++;

      // Identify option index for "Color"
      const options = product.options || [];
      const colorOptionIndex = options.findIndex((opt) => COLOR_OPTION_REGEX.test(opt.name) || /color/i.test(opt.name));
      const colorOptionKey = colorOptionIndex !== -1 ? `option${colorOptionIndex + 1}` : null;

      // Map image_id -> assigned variant color name(s)
      const imageVariantColorMap = {};
      if (product.variants && Array.isArray(product.variants)) {
        for (const variant of product.variants) {
          if (variant.image_id) {
            const colorVal = colorOptionKey ? variant[colorOptionKey] : null;
            if (colorVal) {
              if (!imageVariantColorMap[variant.image_id]) {
                imageVariantColorMap[variant.image_id] = new Set();
              }
              imageVariantColorMap[variant.image_id].add(colorVal.trim());
            }
          }
        }
      }

      const images = product.images || [];

      for (const img of images) {
        totalImagesChecked++;

        // Skip images that already have non-empty alt text
        if (img.alt && String(img.alt).trim() !== '') {
          skippedAlreadyHasAlt++;
          continue;
        }

        // Determine new Alt Text:
        let newAltText = '';

        const assignedColors = imageVariantColorMap[img.id] ? Array.from(imageVariantColorMap[img.id]) : [];
        if (assignedColors.length > 0) {
          newAltText = assignedColors.join(' / ');
        } else {
          // Fallback to product title if not assigned to a specific variant color
          newAltText = product.title;
        }

        updatedImagesCount++;
        actionLog.push({
          productId: product.id,
          productTitle: product.title,
          imageId: img.id,
          src: img.src,
          newAltText,
        });

        console.log(` [${IS_DRY_RUN ? 'DRY-RUN' : 'UPDATING'}] Product: "${product.title}" | Image ID: ${img.id} => Alt: "${newAltText}"`);

        if (!IS_DRY_RUN) {
          await shopifyRequest(`/products/${product.id}/images/${img.id}.json`, {
            method: 'PUT',
            body: JSON.stringify({
              image: {
                id: img.id,
                alt: newAltText,
              },
            }),
          });
        }
      }
    }

    nextUrl = getNextPageUrl(linkHeader);
  }

  console.log('\n====================================================');
  console.log('   SUMMARY REPORT');
  console.log('====================================================');
  console.log(` Total Products Scanned   : ${productsScanned}`);
  console.log(` Total Images Examined    : ${totalImagesChecked}`);
  console.log(` Skipped (Alt Text Exists): ${skippedAlreadyHasAlt}`);
  console.log(` Images ${IS_DRY_RUN ? 'Planned for Update' : 'Updated'}: ${updatedImagesCount}`);
  console.log('====================================================\n');

  if (actionLog.length > 0) {
    console.log('Detailed Log of Updated Images:');
    actionLog.forEach((item, idx) => {
      console.log(`${idx + 1}. [Product ID: ${item.productId}] "${item.productTitle}"`);
      console.log(`   Image ID : ${item.imageId}`);
      console.log(`   New Alt  : "${item.newAltText}"\n`);
    });
  } else {
    console.log('All product images already have valid alt text! No updates needed.');
  }

  if (IS_DRY_RUN && updatedImagesCount > 0) {
    console.log('\n----------------------------------------------------');
    console.log(' NOTE: This was a DRY RUN. No changes were committed to Shopify.');
    console.log(' To apply these updates live to your store, run:');
    console.log('   SHOPIFY_ADMIN_TOKEN=your_token node scripts/fix-missing-alt-text.js --commit');
    console.log('----------------------------------------------------\n');
  }
}

main().catch((err) => {
  console.error('\n Fatal Error encountered during script execution:');
  console.error(err);
  process.exit(1);
});

File: scripts/fix_missing_alt_text.py
#!/usr/bin/env python3
"""
Bulk Fix Missing Alt Text for Shopify Product Images

Usage:
  # Dry run (default):
  SHOPIFY_STORE_DOMAIN=inkitchonline.myshopify.com SHOPIFY_ADMIN_TOKEN=shpat_xxx python scripts/fix_missing_alt_text.py

  # Execute live updates:
  SHOPIFY_STORE_DOMAIN=inkitchonline.myshopify.com SHOPIFY_ADMIN_TOKEN=shpat_xxx python scripts/fix_missing_alt_text.py --commit
"""

import os
import sys
import re
import time
import json
import urllib.request
import urllib.parse
import urllib.error

STORE_DOMAIN = os.getenv("SHOPIFY_STORE_DOMAIN", "inkitchonline.myshopify.com")
ADMIN_TOKEN = os.getenv("SHOPIFY_ADMIN_TOKEN", "")
API_VERSION = os.getenv("SHOPIFY_API_VERSION", "2025-01")
IS_DRY_RUN = "--commit" not in sys.argv and os.getenv("DRY_RUN", "true").lower() != "false"

COLOR_OPTION_REGEX = re.compile(r"^(colou?r|farbe|couleur|coloris|colore|cor|shade|finish|style)$", re.IGNORECASE)

def shopify_request(endpoint_url, data=None, method="GET", retries=5):
    if endpoint_url.startswith("http"):
        url = endpoint_url
    else:
        url = f"https://{STORE_DOMAIN}/admin/api/{API_VERSION}{endpoint_url}"

    headers = {
        "Content-Type": "application/json",
        "X-Shopify-Access-Token": ADMIN_TOKEN
    }

    req_data = json.dumps(data).encode("utf-8") if data else None
    req = urllib.request.Request(url, data=req_data, headers=headers, method=method)

    for attempt in range(1, retries + 1):
        try:
            with urllib.request.urlopen(req) as resp:
                resp_body = resp.read().decode("utf-8")
                link_header = resp.headers.get("Link")
                return json.loads(resp_body), link_header
        except urllib.error.HTTPError as e:
            if e.code == 429:
                retry_after = float(e.headers.get("Retry-After", 2.0))
                print(f" Rate limited (429). Waiting {retry_after}s (attempt {attempt}/{retries})...")
                time.sleep(retry_after + 0.5)
                continue
            else:
                error_body = e.read().decode("utf-8")
                raise Exception(f"Shopify API Error ({e.code} {e.reason}): {error_body}")
        except Exception as e:
            if attempt == retries:
                raise e
            backoff = (2 ** attempt) * 0.5
            print(f" Network error: {e}. Retrying in {backoff}s...")
            time.sleep(backoff)

def parse_next_link(link_header):
    if not link_header:
        return None
    for part in link_header.split(","):
        if 'rel="next"' in part:
            match = re.search(r"<([^>]+)>", part)
            if match:
                return match.group(1)
    return None

def main():
    print("====================================================")
    print("   SHOPIFY BULK ALT-TEXT FIXER (Python)")
    print("====================================================")
    print(f" Store Domain : {STORE_DOMAIN}")
    print(f" API Version  : {API_VERSION}")
    print(f" Mode         : {'DRY RUN (No changes saved)' if IS_DRY_RUN else 'COMMIT (Updating Shopify Admin live!)'}")
    print("====================================================\n")

    if not ADMIN_TOKEN:
        print("ERROR: SHOPIFY_ADMIN_TOKEN environment variable is missing.")
        print("Please set SHOPIFY_ADMIN_TOKEN and run again.\n")
        print("Example:")
        print("  SHOPIFY_ADMIN_TOKEN=shpat_xxxx python scripts/fix_missing_alt_text.py --commit\n")
        sys.exit(1)

    products_scanned = 0
    total_images_checked = 0
    skipped_already_has_alt = 0
    updated_images_count = 0
    action_log = []

    next_url = "/products.json?limit=250"

    while next_url:
        print("Fetching products page...")
        res_data, link_header = shopify_request(next_url)
        products = res_data.get("products", [])

        for product in products:
            products_scanned += 1
            title = product.get("title", "")
            options = product.get("options", [])
            color_opt_index = None

            for idx, opt in enumerate(options):
                opt_name = opt.get("name", "")
                if COLOR_OPTION_REGEX.match(opt_name) or "color" in opt_name.lower():
                    color_opt_index = idx + 1
                    break

            color_option_key = f"option{color_opt_index}" if color_opt_index else None

            # Map image_id -> assigned variant color name(s)
            image_variant_color_map = {}
            for variant in product.get("variants", []):
                img_id = variant.get("image_id")
                if img_id and color_option_key:
                    color_val = variant.get(color_option_key)
                    if color_val:
                        if img_id not in image_variant_color_map:
                            image_variant_color_map[img_id] = set()
                        image_variant_color_map[img_id].add(color_val.strip())

            for img in product.get("images", []):
                total_images_checked += 1
                img_id = img.get("id")
                existing_alt = img.get("alt")

                if existing_alt and str(existing_alt).strip():
                    skipped_already_has_alt += 1
                    continue

                assigned_colors = list(image_variant_color_map.get(img_id, []))
                if assigned_colors:
                    new_alt = " / ".join(assigned_colors)
                else:
                    new_alt = title

                updated_images_count += 1
                action_log.append({
                    "product_id": product.get("id"),
                    "product_title": title,
                    "image_id": img_id,
                    "new_alt": new_alt
                })

                print(f" [{'DRY-RUN' if IS_DRY_RUN else 'UPDATING'}] Product: \"{title}\" | Image ID: {img_id} => Alt: \"{new_alt}\"")

                if not IS_DRY_RUN:
                    shopify_request(
                        f"/products/{product.get('id')}/images/{img_id}.json",
                        data={"image": {"id": img_id, "alt": new_alt}},
                        method="PUT"
                    )

        next_url = parse_next_link(link_header)

    print("\n====================================================")
    print("   SUMMARY REPORT")
    print("====================================================")
    print(f" Total Products Scanned   : {products_scanned}")
    print(f" Total Images Examined    : {total_images_checked}")
    print(f" Skipped (Alt Text Exists): {skipped_already_has_alt}")
    print(f" Images {'Planned for Update' if IS_DRY_RUN else 'Updated'}: {updated_images_count}")
    print("====================================================\n")

    if action_log:
        print("Detailed Log of Updated Images:")
        for idx, item in enumerate(action_log, start=1):
            print(f"{idx}. [Product ID: {item['product_id']}] \"{item['product_title']}\"")
            print(f"   Image ID : {item['image_id']}")
            print(f"   New Alt  : \"{item['new_alt']}\"\n")
    else:
        print("All product images already have valid alt text! No updates needed.")

    if IS_DRY_RUN and updated_images_count > 0:
        print("\n----------------------------------------------------")
        print(" NOTE: This was a DRY RUN. No changes were committed to Shopify.")
        print(" To apply these updates live to your store, run:")
        print("   SHOPIFY_ADMIN_TOKEN=your_token python scripts/fix_missing_alt_text.py --commit")
        print("----------------------------------------------------\n")

if __name__ == "__main__":
    main()
